flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java

```diff
@@ -56,6 +57,10 @@
 
 	protected final SingleInputGate inputGate;
 
+	protected long currentCheckpointId = -1;
+
+	protected ChannelState channelState = ChannelState.CONSUMING;
```

Comment:
I considered discarding these two states in `InputChannel` level to unify reuse the states inside `CheckpointBarrierAligner`.
Regarding the `RemoteInputChannel` case, these two states are used for avoiding unnecessary notification for upstream side if the unblock is triggered by other channels.

One possible option is to pass the boolean array of blocked channels` from `CheckpointBarrierAligner` while calling `notifyCheckpointCompletedOrCanceled`. In `CheckpointedInputGate` and `UnionInputGate` level, they can also  pass the maintained `offsetIndex` to dedicated `SingleInputGate`, then the single gate can judge whether the respective `RemoteInputChannel` is actually blocked or not to notify the upstream side if necessary.

Regarding the `LocalInputChannel` case, we can consider it separately.

Reply:
For the case of `LocalInputChannel`, the current problem is that when the subparition view notifies the data available, then the respective local channel would be added into gate data queue. When the gate pools buffer from such local channel, it should be aware whether this local channel was already blocked by `CheckpointBarrierAligner` or not. If blocked, it should not return this buffer to upper component to avoid caching.

Another option to solve this issue is to avoid adding local channel into gate queue via tagging the blocked state in `ResultSubpartitionView`. We already added this state in `NetworkSequenceViewReader` for remote channel. If we can migrate this state into `ResultSubpartitionView` level, then we can make reuse of this state for both remote and local channels. And I think it should be transparent to do the similar things either in  `NetworkSequenceViewReader` or `ResultSubpartitionView`.